relayer/src/channel.rs

```diff
@@ -420,54 +426,77 @@ impl Channel {
         let a_channel = self
             .src_chain()
             .query_channel(&self.src_port_id(), src_channel_id, Height::zero())
-            .map_err(|_| ChannelError::Failed("Failed to query source chain".into()))?;
+            .map_err(|_| {
+                ChannelError::Failed(format!(
+                    "failed to query source chain {}",
+                    self.src_chain().id()
+                ))
+            })?;
 
         let b_channel = self
             .dst_chain()
             .query_channel(&self.dst_port_id(), dst_channel_id, Height::zero())
-            .map_err(|_| ChannelError::Failed("Failed to query destination chain".into()))?;
+            .map_err(|_| {
+                ChannelError::Failed(format!(
+                    "failed to query destination chain {}",
+                    self.dst_chain().id()
+                ))
+            })?;
 
         match (a_channel.state(), b_channel.state()) {
             (State::Init, State::TryOpen) | (State::TryOpen, State::TryOpen) => {
                 let event = self.flipped().build_chan_open_ack_and_send().map_err(|e| {
-                    error!("Failed ChanAck {:?}: {}", self.a_side, e);
+                    error!("failed ChanAck {:?}: {}", self.a_side, e);
                     e
                 })?;
 
                 info!("done {} => {:#?}\n", self.src_chain().id(), event);
+                // One more step (confirm) left.
+                // Returning error signals that the caller should retry.
+                Err(ChannelError::PartialOpenHandshake(
```

Comment:
I feel that we should not mix up intentional retry with error retry. In this case, if I understand correctly, after the call to `build_chan_open_ack_and_send`, the channel states should progress to the expected state right? So it probably makes sense to rewrite the function into something like follows:

```rust
fn get_channel_states(&self) -> Result<(State, State), ChannelError> { ... }

fn do_chan_open_finalize(&self) -> Result<(), ChannelError> {
  match self.get_channel_states()? {
    (State::Init, State::TryOpen) => {
      let event = self.flipped().build_chan_open_ack_and_send().map_err(...)?;
      match self.get_channel_states()? {
        (State::TryOpen, State::TryOpen) => Ok(()),
        (state1, state2) => Err(...) // channel states did not update to expected states
      }
    }
    (State::TryOpen, State::TryOpen) => {
      // Similar to above, but different expected states
    }
    ...
  }
}
```

Reply:
> I feel that we should not mix up intentional retry with error retry. 

I agree. But I didn't notice any concise/clear way to distinguish between the two here. Could you elaborate on your proposal (taking into account the remark below).

> ... after the call to `build_chan_open_ack_and_send`, the channel states should progress to the expected state right? 

The call to `build_chan_open_confirm_and_send` is the last one (not `..ack..`) which the method `do_chan_open_finalize` must trigger.

Reply:
The idea is that `do_chan_open_finalize` should be a complete function of its own without the need to use `do_chan_open_finalize_with_retry`. The user should have the option to call `do_chan_open_finalize`, and only use `do_chan_open_finalize_with_retry` for additional reliability.

In your comment you write "Returning error signals that the caller should retry". My impression is that it is deliberately returning error so that `retry_with_index` will call `do_chan_open_finalize` again to check if the new state after the call to `build_chan_open_ack_and_send()` matches the other branch. Is my understanding correct?

Reply:
> The user should have the option to call `do_chan_open_finalize`, and only use `do_chan_open_finalize_with_retry` for additional reliability.

Your comment gave me the idea to rename `do_chan_open_finalize` into `do_chan_open_ack_confirm_step`. I think this makes it explicit in the method name that this function is a single step towards opening a channel, and this method expects that it may need to be called multiple times.

420abb54


> My impression is that it is deliberately returning error so that `retry_with_index` will call `do_chan_open_finalize` again

That's correct. The intent is to signal to the caller `retry_with_index` to retry.

The problem with bundling intentional retry with error retry still remains. I can't think of a way to solve that without refactoring this method into two parts, or making the `retry_with_index` block more complicated. But I think this problem is not so concerning anymore .


Please look at the new code and if you can think of a simple way to fix the error bundling issue, I'll give it a try. But I think in its current state (with thorough doc comments) is good.

Reply:
Can we do some logging and see what actually happens when this is being executed? Is it going to retry once and then succeed in the happy path? Or is it going to non-deterministically keep retrying for a few times depending on the block update time?

If the state should be updated right after `build_chan_open_ack_and_send`, then we just need to check the updated the state once, and return error if the state doesn't match so that the error retry logic kicks in.

> I can't think of a way to solve that without refactoring this method into two parts

I do not see a problem of defining helper functions to help structure complicated control flow. Btw FYI you can define non-closure functions and even structs inside a function so that the helper function is only visible within that function. This is more common in languages like Haskell which programmers can use the `where` block to define helper functions without polluting the module namespace.

Reply:
We decided to try to follow-up on this topic in another branch/PR. Specifically, Soares has [some ideas and ongoing work](https://github.com/informalsystems/ibc-rs/compare/adi/1038_fakechain...soares/1038_fakechain) on refactoring the `build_chan_open_ack_and_send` method to make the method's responsibilities and behavior more clear and less bug-prone by splitting this method into smaller helper functions.